Conv Shift Operator

mkliegl opened the pull request:
This closes #4574 .

This is an unoptimized Eigen-based port of the existing code:
- [ConvShiftLayer.cpp](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/gserver/layers/ConvShiftLayer.cpp)
- [conv_shift api](http://doc.paddlepaddle.org/develop/doc/api/v2/config/layer.html#conv-shift)

Examples of possible future optimizations include:
- For M >> log N or N >> log M, it could make sense to use FFT to do circular convolution.
- Instead of a loop over the batch dimension on the outside, it could be faster to do a batchwise operation inside the loop like `out.col(i) = x.col(index) * y.col(j)`. This would make the most sense if we either had column-major storage order or if the batch dimension were last rather than first.

I would probably leave such optimizations to the future when we can profile an actual use case and see whether the improvements are worth it.

@dzhwinter Would you mind taking a look? This is my first PR to this repo - your feedback would be much appreciated!

mkliegl commented:
I think this code does not work for GPU - I will try to fix that. For an overall review, it may be better to wait until that is done. (That said, this code will probably stay about the same for CPU, so I would appreciate feedback on that any time.)

dzhwinter commented on paddle/operators/conv_shift_op.h:
```diff
@@ -0,0 +1,135 @@
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
```
use `using` in the header file violate the google style,  see the detail in [Namespace](https://google.github.io/styleguide/cppguide.html#Namespaces)
> Do not use Namespace aliases at namespace scope in header files except in explicitly marked internal-only namespaces, because anything imported into a namespace in a header file becomes part of the public API exported by that file.

By the way,  I think
```c++
using framework::Tensor;
```
is enough here. 

Some old operators have a wrong guide, sorry for that.

lcy-seso replied:
I just see this comment by chance.

The Google style says: 
> Do not use **Namespace aliases** at namespace scope in header files

I think:
```cpp
using Tensor = framework::Tensor
```
this usage does not violate the above Google style because it is just a type alias, not a namespace alias. But, I admit that using type alias in such a large scope is not a good thing, so you are right...

dzhwinter commented on paddle/operators/conv_shift_op.h:
```diff
@@ -0,0 +1,135 @@
+    size_t y_width = y_dims[1];
+    size_t y_half_width = (y_width - 1) / 2;
+
+    // The below trades code duplication for efficiency (keeping the if
```
good point! 
An if condition inside a three-fold for loop will cost a lot of time.

mkliegl commented:
@dzhwinter @lcy-seso 

Thank you for the feedback! I moved the `using` code you mentioned into the *.cc file and wrote an initial GPU implementation. I also cleaned the code up a little bit. The code is not optimized much but passes the tests. I think it is ready to be merged.

lcy-seso commented:
This PR LGTM, quite clean and well commented~
